src/model_objects/component.rs

```diff
@@ -219,3 +455,323 @@ impl Declarations {
+        assert_eq!(
+            (context.test_comp.clock_usages.get("x").unwrap().edges == context.expected),
+            verdict
+        );
+    }
+
+    #[test_case("Machine",  vec!["L4".to_string()],  true;  "Clock with usage in one invariant")]
+    #[test_case("Machine",  vec!["L6".to_string()],  false; "Clock with usage in one fake invariant")]
+    fn populate_usages_with_invariants(
+        comp_name: &str,
+        expected_locations: Vec<String>,
+        verdict: bool,
+    ) {
+        let mut context = setup(comp_name, expected_locations);
+
+        context.test_comp.populate_usages_with_invariants();
+
+        assert_eq!(
+            (context.test_comp.clock_usages.get("y").unwrap().locations == context.expected),
+            verdict
+        );
+    }
+    // Clock Reduction tests
+
+    /// Test for testing complete clock reduction, by removing redundant clocks and compressing the declarations
+    #[test]
+    fn clock_reduction() {
+        let mut project_loader = JsonProjectLoader::new_loader(PATH, crate::tests::TEST_SETTINGS);
+        project_loader.get_settings_mut().disable_clock_reduction = true;
+        let mut test_comp = project_loader.get_component("Component1").unwrap().clone();
+
+        let expected: HashMap<String, ClockIndex> = HashMap::from([
+            ("x".to_string(), 1),
+            ("y".to_string(), 1),
+            ("z".to_string(), 1),
+            ("i".to_string(), 2),
+        ]);
+
+        // Initializing component clock scopes.
+        test_comp.initialise_clock_usages();
+        test_comp.populate_usages_with_guards();
+        test_comp.populate_usages_with_updates();
+        test_comp.populate_usages_with_invariants();
+
+        // Clock reduction.
+        test_comp
+            .remove_redundant_clocks()
+            .expect("Could not remove redundant clocks.");
+        test_comp.compress_dcls();
+
+        assert_eq!(test_comp.declarations.clocks, expected);
+    }
+
+    #[test]
+    fn remove_redundant_clocks() {
+        let mut project_loader = JsonProjectLoader::new_loader(PATH, crate::tests::TEST_SETTINGS);
+        project_loader.get_settings_mut().disable_clock_reduction = true;
+        let mut test_comp = project_loader.get_component("Component1").unwrap().clone();
+
+        let expected: HashMap<String, ClockIndex> = HashMap::from([
+            ("x".to_string(), 1),
+            ("y".to_string(), 1),
+            ("z".to_string(), 1),
+            ("i".to_string(), 4),
+        ]);
+
+        test_comp.initialise_clock_usages();
+        test_comp.populate_usages_with_guards();
+        test_comp.populate_usages_with_updates();
+        test_comp.populate_usages_with_invariants();
+
+        test_comp
+            .remove_redundant_clocks()
+            .expect("Could not remove redundant clocks.");
+
+        // TODO Test for remapped clocks instead of just if they exist in component
```

Comment:
Just for some consistency
```suggestion
        // TODO: Test for remapped clocks instead of just if they exist in component
```